crypto/src/base58/encodable.rs

```diff
@@ -1,122 +1,108 @@
-    /// to the type.
-    const VERSION_BYTE: u8;
     const MINA_VERSION_BYTE_COUNT: usize = 1;
     const MINA_VERSION_BYTE: u8 = 1;
-    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>>
-    where
-        Self: AsRef<[u8; 32]>,
-    {
+    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>> {
         let mut buf = Vec::with_capacity(32 + Self::MINA_VERSION_BYTE_COUNT);
         for _i in 0..Self::MINA_VERSION_BYTE_COUNT {
             buf.push(Self::MINA_VERSION_BYTE);
         }
-        for &b in self.as_ref() {
-            buf.push(b);
-        }
+        buf.extend(self.to_bytes());
         bs58::encode(buf).with_check_version(Self::VERSION_BYTE)
     }
 
-    fn to_base58_string(&self) -> String
-    where
-        Self: AsRef<[u8; 32]>,
-    {
+    fn to_base58_string(&self) -> String {
         self.to_base58().into_string()
     }
 
-    fn from_base58<I>(i: I) -> Result<Self, bin_prot::error::Error>
+    fn from_base58(i: impl AsRef<[u8]>) -> Result<Self, Error>
     where
-        I: AsRef<[u8]>,
-        Self: From<[u8; 32]>,
+        Self: TryFrom<Vec<u8>>,
+        <Self as TryFrom<std::vec::Vec<u8>>>::Error: std::fmt::Debug,
     {
         let bytes: Vec<u8> = bs58::decode(i)
             .with_check(Some(Self::VERSION_BYTE))
             .into_vec()
-            .map_err(|e| bin_prot::error::Error::Custom {
-                message: format!("{:?}", e),
-            })?;
+            .map_err(|e| Error::OtherError(format!("{:?}", e)))?;
 
         // skip the bs58 version byte and mina bin_prot version bytes
-        let mut b32 = [0; 32];
-        b32.copy_from_slice(&bytes[(1 + Self::MINA_VERSION_BYTE_COUNT)..]);
-        Ok(b32.into())
+        let stored_bytes: Vec<u8> = bytes
+            .into_iter()
+            .skip(1 + Self::MINA_VERSION_BYTE_COUNT)
+            .collect();
+        stored_bytes
+            .try_into()
+            .map_err(|e| Error::OtherError(format!("{:?}", e)))
     }
+
+    // Have to sacrifice perf here in order to share the trait between bin-prot and hash
+    fn to_bytes(&self) -> Vec<u8>;
+}
+
+#[macro_export]
+macro_rules! impl_bs58_for_binprot {
+    ($ty:ty, $expr:expr) => {
+        impl Base58Encodable for $ty {
+            const VERSION_BYTE: u8 = $expr;
+            const MINA_VERSION_BYTE_COUNT: usize = 0;
+
+            fn to_bytes(&self) -> Vec<u8> {
+                let mut buf = Vec::<u8>::new();
+                bin_prot::to_writer(&mut buf, self)
+                    .expect("Failed to serialize struct into binprot format");
+                buf
+            }
+        }
+
+        impl TryFrom<Vec<u8>> for $ty {
+            type Error = bin_prot::error::Error;
+            fn try_from(h: Vec<u8>) -> Result<Self, Self::Error> {
+                bin_prot::from_reader(h.as_slice())
+            }
+        }
+    };
 }
 
 #[macro_export]
-macro_rules! impl_hash_bs58_full {
+macro_rules! impl_bs58_for_hash_full {
```

Comment:
This is kind of a nitpick but this macro isn't just for hashes, it is for any strings of bytes right? This is a recurring source of confusion. To me a hash specifically means 'the output of a hash function'

Reply:
fixed